### PR TITLE
weechat: seperate weechat-unwrapped from wrapper

### PR DESCRIPTION
**Pull request by alyssais**
##### Motivation for this change

If I have a patch I want to apply to weechat, I can't do that with overrideAttrs like I can with almost every other package, because that only applies to the wrapper derivation. For other wrapped packages, one can usually call the wrapper with any version of the derivation, but the weechat derivation didn't expose a wrapper creation function.

Taking inspiration from other packages, particularly Firefox, I extracted the wrapper into its own function, made the default weechat derivation use that, and added weechat-unwrapped.

Now I can add my custom patch like this:

```nix
(wrapWeechat
  (weechat-unwrapped.overrideAttrs (oldAttrs: {
    patches = [
      (fetchpatch {
        url = "https://github.com/weechat/weechat/commit/55767f5f116db3cb56cf85f52aa80feff45b6abf.patch?full_index=1";
        sha256 = "1pkcdsby57diqds1y5hhl0fr4i8j0zax32jb0gqd36siki3lza3d";
      })
    ];
  }))
  { configure =
    { availablePlugins, ... }:
    {
      plugins = with availablePlugins; [
        (python.withPackages (packages: with packages; [ potr websocket_client ]))
      ];
    };
  })
```

##### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [x] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

There is a small backward incompatibility here: previously, it was possible to get an unwrapped weechat like this:

```nix
weechat.override { configure = null; }
```

This didn't seem too important to keep around since it was also possible to get an unwrapped weechat in a much more obvious way:

```nix
weechat.unwrapped
```

I could probably make it so that the first way still worked, if that behavior turns out to really have been important.

### CODE REVIEW

**Comment by Mic92** [comment]
cc @lheckemann

**Comment by lheckemann** [comment]
Not tested or read this in detail but :+1:  on the concept!

**Comment by matthewbauer** [comment]
Sorry for the late response! It looks like there is a merge conflict.

One opinion I have on wrappers is that we should avoid them unless absolutely necessary. Packages in all-packages.nix should just be plain packages - no configuration should be required. Your PR probably improves the wrapper situation but I am wondering if there is a better way to do it.

Could we just set WEECHAT_EXTRA_LIBDIR to $HOME/.nix-profile/lib/weechat (and maybe add some more search paths if it's allowed)? That would mean you could just install each plugin like you would through nix-env. That seems easier than just wrapping to me.

I'm not a maintainer of weechat or anything, just interested in improving some of our wrapper situations.

**Comment by lheckemann** [comment]
@matthewbauer I'm not sure I like the idea of not using a wrapper. It's a lot more versatile than using `.nix-profile`, allowing:
- testing different plugin sets with `nix-shell` or `nix run` without having to mutate your profile
- installing it in the system profile without having to extend the patch to support multiple search paths (though that is admittedly a limitation which would be nice to remove)
- creating a systemd user service which runs it in tmux without installing it in the profile
- using profiles other than `~/.nix-profile`

Conceptually, it makes the dependencies between weechat and its plugins explicit, and minimising implicit dependencies is one of the things I like so much about nix.

**Comment by matthewbauer** [comment]
I suppose it's all okay if there's documentation (which looks well written for weechat!) My main beef with wrappers is that you almost always have to end up digging into the source to figure out how to configure them.

So the original reason for this was to avoid large closures, correct? I assume most distros would just ship with every possible language weechat supports but that could be huge.

**Comment by lheckemann** [comment]
>So the original reason for this was to avoid large closures, correct?

Yes.

>I assume most distros would just ship with every possible language weechat supports but that could be huge.


Most distros would build separate binary plugin packages from a single source package and install them into a single shared namespace like `/usr/lib/weechat`. But we can't do that in nix :D

**Comment by lheckemann** [comment]
I've taken the liberty of merging current master in here. If someone could do a sanity check on this and merge (or if you don't have commit access just let me know) that would be great :)

**Comment by lheckemann** [comment]
Oh no, I think I did it wrong :|

**Comment by lheckemann** [comment]
Force-pushed my second attempt. Sorry about the chaos!

@kenogo I think you're on the wrong PR?

**Comment by kenogo** [comment]
@lheckemann Excuse me? I don't see any interaction of myself here

**Comment by lheckemann** [comment]
Oops, never mind, sorry. The comment seems to have appeared as a result of the github mess with the merge.

**Comment by alyssais** [comment]
Now that I know more about Nixpkgs (this was one of my first PRs), I'm not sure this is the way to do it. I think we could just use `makeWrapper` for this, and that would remove the headache of there being two derivations.

**Comment by lheckemann** [comment]
I don't see how that would avoid any of the stuff we're already doing — we need to create the plugins dir one way or another and makeWrapper won't really make this any cleaner until we have a nix-language wrapper for it (something I've been meaning to do for a long time…)

**Comment by alyssais** [comment]
Could just use `cat` in the `postInstall` too – other derivations do this. The advantage would be that I could use `overrideAttrs` without doing anything special, which was the original motivation for this PR.

**Comment by lheckemann** [comment]
That sounds like it would prevent changing the plugin set without rebuilding weechat (at which point you could also just not separate the outputs and disable the plugins at build time), or am I misunderstanding something?

**Comment by alyssais** [comment]
Oh, I see. Yes, that's correct. I think that would be a feature, though – would mean you didn't get a bloated closure if you weren't using some of the plugins.

**Comment by lheckemann** [comment]
That's why I split weechat's outputs and wrote the wrapper mechanism in the first place — so you can use plugins of your choice without bloating the closure while still having the binary cache available.

**Comment by alyssais** [comment]
So you're saying that you can choose either:

- A bloated closure with all the optional dependencies, that pulls from the binary cache. You can then customise the plugins with `weechat.override { configure = ...; }`.
- A non-bloated closure with only the optional dependencies you specify, that has to build from source because your custom closure won't be cached. You'd customise the plugins with `weechat.override { whateverSupport = ...; }`

So the user has to basically make a tradeoff between closure size and building from source. Have I understood that correctly?

**Comment by lheckemann** [comment]
Yes, that was the situation before the wrapper. The wrapper allows both using the binary cache and having a smaller closure.

**Comment by zimbatm** [comment]
Are the plugins compiled against the weechat source code? If not you might also want to adopt the approach we took in the terraform package.

**Comment by lheckemann** [comment]
Not sure if they are actually, but they're part of the weechat source distribution so IMHO it makes sense not to split them into separate derivations.

**Comment by alyssais** [comment]
> The wrapper allows both using the binary cache and having a smaller closure.

How? Surely when using the wrapper, the wrapped weechat pulled from the binary cache has Ruby, Lua, etc in its closure even though they aren’t used?

**Comment by lheckemann** [comment]
See also #25274 for the original introduction of the wrapper approach.

**Comment by alyssais** [comment]
Oh, my apologies. I didn’t notice the multiple outputs.

File: pkgs/top-level/all-packages.nix [hunk 1]
```diff
@@ -19712,12 +19712,22 @@ with pkgs;
...
     guile = guile_2_0;
   };
 
+  weechat = wrapWeechat weechat-unwrapped { };
+
+  weechat-matrix-bridge = callPackage ../applications/networking/instant-messengers/weechat-matrix-bridge {
+    inherit (luaPackages) cjson;
+  };
+
+  weechat-xmpp = callPackage ../applications/networking/instant-messengers/weechat-xmpp {};
```
Comment (alyssais) [inline pkgs/top-level/all-packages.nix]: This isn’t supposed to be in the diff, right?

Comment (lheckemann) [reply pkgs/top-level/all-packages.nix]: Indeed, good catch! I didn't realise these had moved into weechatScripts. Thanks!

**Comment by alyssais** [comment]
Cool! Thanks for the explanations @lheckemann. I'm happy with this again. Anything else needing to be done before merging?

**Comment by lheckemann** [comment]
Since `weechat.override {configure = …;}` still works the same, bar the `null` value which wasn't documented, I don't see any documentation changes necessary. Thanks for the PR and sorry it took so long to go through :)

File: pkgs/applications/networking/irc/weechat/wrapper.nix [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+      for plugin in ${lib.concatMapStringsSep " " (p: p.pluginFile) plugins} ; do
+        ln -s $plugin $out/plugins
+      done
+    '';
+
+  in (writeScriptBin "weechat" ''
+    #!${stdenv.shell}
+    export WEECHAT_EXTRA_LIBDIR=${pluginsDir}
+    ${lib.concatMapStringsSep "\n" (p: lib.optionalString (p ? extraEnv) p.extraEnv) plugins}
+    exec ${weechat}/bin/weechat "$@"
```
Comment (Ma27) [inline pkgs/applications/networking/irc/weechat/wrapper.nix]: it seems as after the merge the `init` support with `--run-command` got lost @lheckemann @alyssais 

was this intentional?

Comment (alyssais) [reply pkgs/applications/networking/irc/weechat/wrapper.nix]: nope

Comment (alyssais) [reply pkgs/applications/networking/irc/weechat/wrapper.nix]: https://github.com/NixOS/nixpkgs/pull/51571

**Pull request merged**